feat: Initial copy-paste circuit

Kelketek opened the pull request:
This PR creates the initial scaffolding for the section_to_course application.

**JIRA tickets**: https://tasks.opencraft.com/browse/BB-7295

**Dependencies**: https://github.com/openedx/edx-platform/pull/32058

**Testing instructions**:

To set up the app:

1. Check out the edx-platform branch as linked in the Dependencies section
2. In the `src` directory of the devstack, clone this repository and checkout this branch.
3. In the `devstack` repo root, run `make dev.shell.studio`
4. `cd /edx/src/section_to_course`
5. `pip install -e .`

To run the unit tests:

7. From the `devstack` repo root, run `make dev.shell.studio`
8. `DJANGO_SETTINGS_MODULE=cms.envs.test pytest --pyargs section_to_course --rootdir cms`

To test the functionality manually:

1. Create a course (or find an existing one) and pick a section from it. Derive the usage key string (this can be done, for instance, by duplicating a block and checking the network request in the devtools to see what the source block string is)
2. From the `devstack` repo root, run `make dev.shell.studio`
3. Run the management command to perform the copy. The command is `section_to_course` and should be followed by the destination course and the source section's usage key. For example, `./manage.py cms section_to_course course-v1:Beeper+B33P+BOOP block-v1:edX+DemoX+Demo_Course+type@chapter+block@graded_interactions edx`.

**Author notes and concerns**:

The cookiecutter we derived from had a lot of nice CI functionalities, not to mention linting. I had to discard all of it since we rely on the edx-platform environment and can't (readily) import that into the GitHub CI. So, please be attentive to code quality when reviewing.

Agrendalath commented on section_to_course/utils.py:
```diff
@@ -0,0 +1,60 @@
+    destination_course_key = CourseKey.from_string(destination_course_id)
+    destination_course = store.get_course(destination_course_key)
+    block_usage_key = BlockUsageLocator.from_string(block_id)
+    block_key = BlockKey(block_usage_key.block_type, block_usage_key.block_id)
```
Would it make sense to change the `derived_key` function to accept `BlockUsageLocator` instead of the `BlockKey`? The `_copy_from_template` method in the `edx-platform` already has the `BlockUsageLocator` (and we have it here as the `block_usage_key`), so it doesn't make too much sense to have this additional conversion that only creates a `namedtuple`.

Kelketek replied:
I like the symmetry of putting a BlockKey in and getting a derived BlockKey out. It feels cleaner, even though we would save a few lines.

Agrendalath commented on section_to_course/utils.py:
```diff
@@ -0,0 +1,60 @@
+            update_from_source(source_block=block, destination_block=dest_block, user_id=user.id)
+        else:
+            dest_block_location = duplicate_block(
+                parent_usage_key=destination_course.location,
```
```suggestion
                parent_usage_key=block_usage_key.context_key,
```

Kelketek replied:
Nope, those aren't the same thing.

Agrendalath commented on section_to_course/management/commands/section_to_course.py:
```diff
@@ -0,0 +1,32 @@
+
+    def add_arguments(self, parser):
+        parser.add_argument('destination_course_id', type=str)
+        parser.add_argument('section_id', type=str)
```
Nit: `source_section_id` would be a more verbose name.

Agrendalath commented on section_to_course/models.py:
```diff
@@ -0,0 +1,33 @@
+    destination_course_id = CourseKeyField(max_length=255, db_index=True, null=False, blank=False)
+    source_section_id = UsageKeyField(max_length=255, db_index=True, null=False, blank=False)
+    destination_section_id = UsageKeyField(max_length=255, db_index=True, null=False, blank=False)
+    last_refresh = models.DateTimeField(null=True, blank=True)
```
This is a `TimeStampedModel` with the `modified` date. This is redundant.

Kelketek replied:
@Agrendalath I'm not sure it will always be. There is a strong potential later on in the project to annotate this model with more metadata, in which case the rest of the fields will be covered by 'modified' rather than 'last_refresh'.

Agrendalath commented on section_to_course/tests/test_utils.py:
```diff
@@ -0,0 +1,48 @@
+from common.djangoapps.student.tests.factories import UserFactory
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
+from xmodule.modulestore.tests.factories import CourseFactory, BlockFactory
```
You know this, but let's make it visible - these are integration tests, so we should add a TODO to GitHub Actions to set up running them properly in the CI. Running tests manually is not maintainable. This will be especially important when upgrading the platform to Olive/Palm.

If we keep the `upgrade-python-requirements.yml`, we can ensure that the compatibility with the `master` branch of `edx-platform` is tested every week to catch any breaking changes in time.

Kelketek commented on section_to_course/management/commands/tests/test_selection_to_course.py:
```diff
@@ -0,0 +1,4 @@
+"""
+Tests the SectionToCourse management command
```
Whoops! Forgot to actually add the test. I'll do so :)

Kelketek commented on section_to_course/models.py:
```diff
@@ -0,0 +1,33 @@
+    """
+
+    source_course_id = CourseKeyField(max_length=255, db_index=True, null=False, blank=False)
+    destination_course_id = CourseKeyField(max_length=255, db_index=True, null=False, blank=False)
```
@Agrendalath I think both of these fields are going to make more sense when we start writing the admin views and are going to need to look up the values by the source or destination course for some operations to avoid accidental duplication, for instance.

Kelketek commented on section_to_course/utils.py:
```diff
@@ -0,0 +1,60 @@
+        source_course_id=block_usage_key.course_key,
+        destination_course_id=destination_course_key,
+        source_section_id=block_usage_key,
+        destination_section_id=dest_block.location,
```
Sure!

Kelketek commented on setup.py:
```diff
@@ -0,0 +1,138 @@
+        exclude=["*tests"],
+    ),
+    entry_points={
+        "lms.djangoapp": [
+            "section_to_course = section_to_course.apps:SectionToCourseConfig",
+        ],
```
No, I suppose we don't.

Kelketek commented:
@Agrendalath This is ready for another look!

Agrendalath commented on README.rst:
```diff
@@ -0,0 +1,70 @@
+Then, in the `src` directory of your devstack, run:
+
+.. code-block::
+  git clone git@github.com:open-craft/section-to-course.git
```
> These points apply to the whole file.

Please take a look at the next few lines of the README.

Agrendalath commented on requirements/test.in:
```diff
@@ -0,0 +1,9 @@
+pytest-cov                # pytest extension for code coverage statistics
+pytest-django             # pytest extension for better Django support
+code-annotations          # provides commands used by the pii_check make target.
+git+https://github.com/openedx/opaque-keys.git@2.3.0#egg=opaque-keys==2.3.0
```
This is not longer needed in `test.in`.

Agrendalath commented on MANIFEST.in:
```diff
@@ -1,4 +1,3 @@
-include CHANGELOG.rst
```
^

Agrendalath commented on .github/workflows/upgrade-python-requirements.yml:
```diff
@@ -1,24 +0,0 @@
-       requirements_bot_github_email: ${{ secrets.REQUIREMENTS_BOT_GITHUB_EMAIL }}
-       edx_smtp_username: ${{ secrets.EDX_SMTP_USERNAME }}
-       edx_smtp_password: ${{ secrets.EDX_SMTP_PASSWORD }}
-    uses: edx/.github/.github/workflows/upgrade-python-requirements.yml@master
```
^

Agrendalath commented on .coveragerc:
```diff
@@ -1,10 +0,0 @@
-    *migrations*
-    *admin.py
-    *static*
-    *templates*
```
^

Agrendalath commented on codecov.yml:
```diff
@@ -1,12 +0,0 @@
-        enabled: yes
-        target: 100%
-
-comment: false
```
^

Agrendalath commented on section_to_course/management/commands/section_to_course.py:
```diff
@@ -0,0 +1,54 @@
+    help = 'Converts a section into a course'
+
+    def add_arguments(self, parser):
+        parser.add_argument('destination_course_id', type=str)
+        parser.add_argument('source_section_id', type=str)
```
Let's use the `source` before the `destination` in the management command. It is more intuitive (same as [here](https://github.com/open-craft/section-to-course/pull/1#discussion_r1168494486)).

Kelketek commented:
> When I clone the section into a new course, the subsections are not displayed on the Course Outline page. Do we need to refresh learning sequences manually to get it working? When I go to the `[/admin/learning_sequences/coursecontext/](http://localhost:18000/admin/learning_sequences/coursecontext/)` page, I don't see any sequences in the "Debug Details".
> 
> ![image](https://user-images.githubusercontent.com/3189670/233030036-bb05ce99-35a2-4a25-a270-1bff538b802d.png)
> 
> Also, we will need to backport the upstream PR to our Nutmeg branch and test it there before we merge this, right?

@Agrendalath I'm not able to reproduce this issue-- when I copy the section over and then refresh the learning MFE page, I see the new section in the outline. Can you give me step-by-step instructions to reproduce?

As for backporting the upstream branch, probably! We can make sure to clear that out first. I'll start the next task based on this branch while we wait on that merge. If it takes long enough we'll just put the commit into our Nutmeg branch while we wait.

Or, if you think it would be better, I can create the PR against our nutmeg branch now.

I've addressed the remaining notes-- could you take one more look?

Agrendalath commented:
@Kelketek,
> Can you give me step-by-step instructions to reproduce?

I just followed your testing instructions for the first section of the demo course. It might be something with my devstack.

> I'll start the next task based on this branch while we wait on that merge.

What are we waiting for? I already left my +1 and asked you in Jira if you want me to merge it now or wait for the Nutmeg testing.

> Or, if you think it would be better, I can create the PR against our nutmeg branch now.

Yeah, we will deploy this to a Nutmeg instance, so let's do it now. We also won't need to worry about the Learning Sequences while rendering the Course Outline page.

> I've addressed the remaining notes-- could you take one more look?

I'm still seeing a couple of unresolved discussions. They are marked as "outdated", so you may need to open them manually. They are still valid, though.

Kelketek commented:
> What are we waiting for? I already left my +1 and asked you in Jira if you want me to merge it now or wait for the Nutmeg testing.

Sorry, forgot you had merge rights on edx-platform. I'll create the MR against our Nutmeg branch.

> I'm still seeing a couple of unresolved discussions (https://github.com/open-craft/section-to-course/pull/1#discussion_r1170975676). They are marked as "outdated", so you may need to open them manually. They are still valid, though.

I've made a push to fix up the README. I didn't see any other outstanding discussions than the one you mentioned, though. If there are more, can you ping me on them?

bradenmacdonald commented:
Damn, the sheer amount of boilerplate required in a new django app for Open edX is truly gross.

Kelketek commented:
@Agrendalath I went to backport the changes to Nutmeg and found that the structure/location of items in the code changed much more than I had anticipated. The most recent changes should allow the app to be installed and used in both versions while also allowing the quality tests to work.

Kelketek commented:
@Agrendalath The changes requested by Braden upstream are pretty involved, but I'm expecting the end result will have the same data structures. So, I don't think we need to hold up this PR for it if we juggle things a bit. Here's how I propose we move forward so the project doesn't fall behind but we still get all adjustments made:

1.We'll install and run this based on the [backported PR to nutmeg](https://github.com/open-craft/edx-platform/pull/532).
2. I'll get the admin views in the remaining time this sprint.
3. The sprint after this I have the discovery for programs. I can do that alongside addressing Braden's notes on the upstream PR. If I find time to work on them earlier, I will.

I've added a note into the README about the need to use our Nutmeg branch for now.

bradenmacdonald commented:
@Kelketek I didn't realize any of my feedback would be "pretty involved", so feel free to discuss it more on that upstream PR first if you think it can save you some work.

Agrendalath commented:
@Kelketek,

> I didn't see any other outstanding discussions than the one you mentioned, though. If there are more, can you ping me on them?

I see the following discussions:
1. https://github.com/open-craft/section-to-course/pull/1#discussion_r1168444436
2. https://github.com/open-craft/section-to-course/pull/1#discussion_r1168445136
3. https://github.com/open-craft/section-to-course/pull/1#discussion_r1168445801
4. https://github.com/open-craft/section-to-course/pull/1#discussion_r1168446446

> 1.We'll install and run this based on the https://github.com/open-craft/edx-platform/pull/532.

Is this step ready for review?

Kelketek commented:
Thanks, @Agrendalath . I'm not sure why they weren't appearing for me earlier. I've added the workflows back, and the docs, which I've updated. Of course as you know, most of these will fail at the moment until Sathis completes the CI changes needed.

> Is this step ready for review?

Yes.

Kelketek commented:
> @Kelketek, by https://github.com/open-craft/section-to-course/pull/1#discussion_r1168444436, I meant restoring the CHANGELOG.rst, not the whole docs directory. Having "Check the docs directory <docs/>_" that points to a plain directory without a clear table of content is simply ugly. Having the docs directory for this repo is overkill at this point - a simple README with a clear description is enough.

I know you didn't mean this-- I had elected to pull it back in since a bunch of small snippets were referring to the docs so I put it back in, since it seemed very 'may as well'. But you're right, it's verbose. I've re-removed it.

> Did I miss something? What changes are you referring to?

Possibly. Here's a recap: Back in Columbia I had spoken with Braden asking them if there was any current CI solution to apps like this, which rely on platform internals to run. His answer was that he knew of none and it's a continued issue. This is why I removed most of the workflows. Once I submitted this PR though, you mentioned you want us to find a way to get CI running anyway. But that's a sizable project, [so I scheduled a follow-up ticket](https://tasks.opencraft.com/browse/BB-7366), which is currently assigned to Sathis.

Kelketek commented:
Great. @Agrendalath was there anything else you needed addressed on this PR? Did we want to move forward with merging it in anticipation of running it with the Nutmeg branch?

Kelketek commented:
@Agrendalath Ping! See above question. :)

Agrendalath commented on README.rst:
```diff
@@ -0,0 +1,42 @@
+
+Allows course authors to factor sections from Open edX courses into their own new course.
+
+Check the `docs directory <docs/>`_ for details on how to set up and configure this application.
```
I suppose you forgot to add these back.

Agrendalath reviewed (approved):
See my review comment before merging.

:+1:

* [x] I tested this: followed the testing instructions
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation: not yet ⚠️ 
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

Kelketek commented:
@Agrendalath Went ahead and added the instructions to the README. If you find any issue with them, I'll go ahead and fix it in https://github.com/open-craft/section-to-course/pull/2 . Merging now.

Kelketek commented:
@Agrendalath I realized after I merged this that I forgot to squash, so since we haven't deployed yet I went ahead and did so on master. Not something I anticipate being able to do again. cc @sathiscode
